File: sdk/typescript/src/scanners/UnifiedScanner.ts
// ===========================================================================
// OPT-003: Unified Content Scanner
// Replaces sequential per-scanner regex evaluation with a single-pass
// Aho-Corasick automaton. All scanner patterns compile into one automaton.
// Hard-before-soft ordering: hard-severity match aborts immediately.
// ===========================================================================

import { AhoCorasick, type AhoCorasickMatch } from "./AhoCorasick";
import { extractLiterals } from "./LiteralExtractor";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface ScannerConfig {
  /** Unique scanner identifier (e.g., "pii", "injection", "secrets") */
  scannerId: string;
  /** Human-readable label */
  label: string;
  /** Patterns with severity classification */
  patterns: ScannerPattern[];
}

export interface ScannerPattern {
  /** Unique pattern identifier */
  patternId: string;
  /** The regex pattern to match */
  regex: RegExp;
  /** hard = block immediately, soft = warn */
  severity: "hard" | "soft";
}

export interface ScanResult {
  /** Source scanner ID */
  scannerId: string;
  /** Pattern ID within the scanner */
  patternId: string;
  /** hard or soft severity */
  severity: "hard" | "soft";
  /** Match location in the payload */
  match: {
    start: number;
    end: number;
    text: string;
  };
  /** Human-readable scanner name */
  scannerLabel: string;
}

interface PatternMapping {
  scannerId: string;
  patternId: string;
  severity: "hard" | "soft";
  regex: RegExp;
  scannerLabel: string;
  /** Indices into the Aho-Corasick pattern array for this pattern's literals */
  acIndices: number[];
  /** Whether this pattern has literals for pre-filtering */
  hasLiterals: boolean;
}

// ---------------------------------------------------------------------------
// Unified Scanner
// ---------------------------------------------------------------------------

export class UnifiedScanner {
  private readonly hardAutomaton: AhoCorasick;
  private readonly softAutomaton: AhoCorasick;
  private readonly hardMappings: PatternMapping[] = [];
  private readonly softMappings: PatternMapping[] = [];
  private readonly hardLiteralToPatterns: Map<number, number[]> = new Map();
  private readonly softLiteralToPatterns: Map<number, number[]> = new Map();
  private readonly directRegexPatterns: PatternMapping[] = [];

  constructor(scannerConfigs: ScannerConfig[]) {
    const hardLiterals: string[] = [];
    const softLiterals: string[] = [];

    for (const config of scannerConfigs) {
      for (const pattern of config.patterns) {
        const literals = extractLiterals(pattern.regex);

        const mapping: PatternMapping = {
          scannerId: config.scannerId,
          patternId: pattern.patternId,
          severity: pattern.severity,
          regex: pattern.regex,
          scannerLabel: config.label,
          acIndices: [],
          hasLiterals: literals.length > 0,
        };

        if (literals.length === 0) {
          // No extractable literals: fall back to direct regex
          this.directRegexPatterns.push(mapping);
          continue;
        }

        if (pattern.severity === "hard") {
          const mappingIdx = this.hardMappings.length;
          this.hardMappings.push(mapping);

          for (const literal of literals) {
            const acIdx = hardLiterals.length;
            hardLiterals.push(literal);
            mapping.acIndices.push(acIdx);

            // Map AC index -> pattern mapping indices
            const existing = this.hardLiteralToPatterns.get(acIdx);
            if (existing) {
              existing.push(mappingIdx);
            } else {
              this.hardLiteralToPatterns.set(acIdx, [mappingIdx]);
            }
          }
        } else {
          const mappingIdx = this.softMappings.length;
          this.softMappings.push(mapping);

          for (const literal of literals) {
            const acIdx = softLiterals.length;
            softLiterals.push(literal);
            mapping.acIndices.push(acIdx);

            const existing = this.softLiteralToPatterns.get(acIdx);
            if (existing) {
              existing.push(mappingIdx);
            } else {
              this.softLiteralToPatterns.set(acIdx, [mappingIdx]);
            }
          }
        }
      }
    }

    this.hardAutomaton = new AhoCorasick(hardLiterals);
    this.softAutomaton = new AhoCorasick(softLiterals);
  }

  /**
   * Scan a payload string for all matching patterns.
   *
   * Phase 1: Run hard-severity automaton. If any hard literal matches and
   *          confirms with original regex, return immediately (early abort).
   * Phase 2: Run soft-severity automaton. Confirm matches with original regex.
   *          Return all confirmed findings.
   */
  scan(payload: string): ScanResult[] {
    // Phase 1: Hard-severity patterns
    const hardResult = this.scanWithAutomaton(payload, this.hardAutomaton, this.hardMappings, this.hardLiteralToPatterns);
    if (hardResult.length > 0) {
      // Hard abort: return first hard finding immediately
      return [hardResult[0]];
    }

    // Also check direct-regex hard patterns (no literals)
    for (const mapping of this.directRegexPatterns) {
      if (mapping.severity !== "hard") continue;
      const directResult = this.confirmDirectRegex(payload, mapping);
      if (directResult) return [directResult];
    }

    // Phase 2: Soft-severity patterns
    const results: ScanResult[] = [];

    const softResults = this.scanWithAutomaton(payload, this.softAutomaton, this.softMappings, this.softLiteralToPatterns);
    results.push(...softResults);

    // Direct-regex soft patterns
    for (const mapping of this.directRegexPatterns) {
      if (mapping.severity !== "soft") continue;
      const directResult = this.confirmDirectRegex(payload, mapping);
      if (directResult) results.push(directResult);
    }

    return results;
  }

  /**
   * Scan a streaming payload. Yields results per chunk.
   * Aborts on first hard finding (mid-stream abort).
   */
  async *scanStreaming(chunks: AsyncIterable<string>): AsyncGenerator<ScanResult> {
    let fullPayload = "";

    for await (const chunk of chunks) {
      fullPayload += chunk;

      // Check hard patterns on accumulated payload
      const hardResults = this.scanWithAutomaton(
        fullPayload, this.hardAutomaton, this.hardMappings, this.hardLiteralToPatterns,
      );

      if (hardResults.length > 0) {
        yield hardResults[0];
        return; // Hard abort
      }

      // Check direct-regex hard patterns
      for (const mapping of this.directRegexPatterns) {
        if (mapping.severity !== "hard") continue;
        const directResult = this.confirmDirectRegex(fullPayload, mapping);
        if (directResult) {
          yield directResult;
          return; // Hard abort
        }
      }

      // Yield soft results for this chunk (deduplicated by tracking payload length)
      const softResults = this.scanWithAutomaton(
        fullPayload, this.softAutomaton, this.softMappings, this.softLiteralToPatterns,
      );
      for (const result of softResults) {
        yield result;
      }
    }

    // Final: direct-regex soft patterns on full payload
    for (const mapping of this.directRegexPatterns) {
      if (mapping.severity !== "soft") continue;
      const directResult = this.confirmDirectRegex(fullPayload, mapping);
      if (directResult) yield directResult;
    }
  }

  // -----------------------------------------------------------------------
  // Internal
  // -----------------------------------------------------------------------

  /**
   * Run an Aho-Corasick automaton against the payload, then confirm each
   * candidate with the original regex.
   */
  private scanWithAutomaton(
    payload: string,
    automaton: AhoCorasick,
    mappings: PatternMapping[],
    literalToPatterns: Map<number, number[]>,
  ): ScanResult[] {
    const results: ScanResult[] = [];
    const acMatches = automaton.search(payload);

    // Collect candidate pattern indices from AC matches
    const candidatePatterns = new Set<number>();
    for (const match of acMatches) {
      const patternIndices = literalToPatterns.get(match.patternIndex);
      if (patternIndices) {
        for (const idx of patternIndices) {
          candidatePatterns.add(idx);
        }
      }
    }

    // Confirm each candidate with original regex
    for (const patIdx of candidatePatterns) {
      const mapping = mappings[patIdx];
      const regexResult = this.confirmRegex(payload, mapping);
      if (regexResult) {
        results.push(regexResult);
      }
    }

    return results;
  }

  /**
   * Confirm a pattern match with the original regex.
   */
  private confirmRegex(payload: string, mapping: PatternMapping): ScanResult | null {
    // Create a fresh regex to reset lastIndex
    const regex = new RegExp(mapping.regex.source, mapping.regex.flags.replace("g", ""));
    const match = regex.exec(payload);
    if (!match) return null;

    return {
      scannerId: mapping.scannerId,
      patternId: mapping.patternId,
      severity: mapping.severity,
      match: {
        start: match.index,
        end: match.index + match[0].length,
        text: match[0],
      },
      scannerLabel: mapping.scannerLabel,
    };
  }

  /**
   * Direct regex evaluation for patterns without extractable literals.
   */
  private confirmDirectRegex(payload: string, mapping: PatternMapping): ScanResult | null {
    return this.confirmRegex(payload, mapping);
  }
}

File: sdk/python/dynaep/scanners/unified_scanner.py
"""
OPT-003: Unified Content Scanner (Python)
Replaces sequential per-scanner regex evaluation with single-pass
Aho-Corasick automaton. Hard-before-soft ordering with immediate abort.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import AsyncIterator, Iterator, Optional

from dynaep.scanners.aho_corasick import AhoCorasick


# ---------------------------------------------------------------------------
# Types
# ---------------------------------------------------------------------------

@dataclass
class ScannerPattern:
    pattern_id: str
    regex: re.Pattern[str]
    severity: str  # "hard" | "soft"


@dataclass
class ScannerConfig:
    scanner_id: str
    label: str
    patterns: list[ScannerPattern] = field(default_factory=list)


@dataclass
class ScanResult:
    scanner_id: str
    pattern_id: str
    severity: str
    match_start: int
    match_end: int
    match_text: str
    scanner_label: str


# ---------------------------------------------------------------------------
# Literal Extractor
# ---------------------------------------------------------------------------

_META_CHARS = set(".*+?()[]{}|^$\\")


def extract_literals(regex: re.Pattern[str]) -> list[str]:
    """Extract longest contiguous literal substrings from a regex."""
    source = regex.pattern
    flags = regex.flags
    case_insensitive = bool(flags & re.IGNORECASE)

    literals: list[str] = []
    current = ""
    i = 0

    _shortcuts = set("dDwWsSbBnrtfv0")

    def flush():
        nonlocal current
        if current:
            literals.append(current)
            current = ""

    while i < len(source):
        ch = source[i]

        if ch == "\\":
            if i + 1 < len(source):
                nxt = source[i + 1]
                if nxt not in _shortcuts:
                    current += nxt
                    i += 2
                    continue
                flush()
                i += 2
                continue
            flush()
            i += 1
            continue

        if ch == "[":
            flush()
            depth = 1
            i += 1
            while i < len(source) and depth > 0:
                if source[i] == "\\":
                    i += 2
                    continue
                if source[i] == "]":
                    depth -= 1
                i += 1
            continue

        if ch in ("(", ")", "|", ".", "^", "$"):
            flush()
            i += 1
            continue

        if ch in ("*", "+", "?"):
            if current:
                current = current[:-1]
                flush()
            i += 1
            if i < len(source) and source[i] == "?":
                i += 1
            continue

        if ch == "{":
            close = source.find("}", i)
            if close > i and re.match(r"^\d+(,\d*)?$", source[i + 1:close]):
                if current:
                    current = current[:-1]
                    flush()
                i = close + 1
                if i < len(source) and source[i] == "?":
                    i += 1
                continue

        current += ch
        i += 1

    flush()

    filtered = [l for l in literals if len(l) >= 2]
    if not filtered and literals:
        filtered = literals

    if case_insensitive:
        filtered = [l.lower() for l in filtered]

    return filtered


# ---------------------------------------------------------------------------
# Pattern Mapping
# ---------------------------------------------------------------------------

@dataclass
class _PatternMapping:
    scanner_id: str
    pattern_id: str
    severity: str
    regex: re.Pattern[str]
    scanner_label: str
    ac_indices: list[int] = field(default_factory=list)
    has_literals: bool = True


# ---------------------------------------------------------------------------
# Unified Scanner
# ---------------------------------------------------------------------------

class UnifiedScanner:
    """
    Single-pass multi-pattern content scanner.
    Phase 1: Hard-severity automaton (abort on first match).
    Phase 2: Soft-severity automaton (return all findings).
    """

    def __init__(self, scanner_configs: list[ScannerConfig]):
        hard_literals: list[str] = []
        soft_literals: list[str] = []
        self._hard_mappings: list[_PatternMapping] = []
        self._soft_mappings: list[_PatternMapping] = []
        self._hard_lit_to_pat: dict[int, list[int]] = {}
        self._soft_lit_to_pat: dict[int, list[int]] = {}
        self._direct_patterns: list[_PatternMapping] = []

        for config in scanner_configs:
            for pattern in config.patterns:
                literals = extract_literals(pattern.regex)

                mapping = _PatternMapping(
                    scanner_id=config.scanner_id,
                    pattern_id=pattern.pattern_id,
                    severity=pattern.severity,
                    regex=pattern.regex,
                    scanner_label=config.label,
                    has_literals=len(literals) > 0,
                )

                if not literals:
                    self._direct_patterns.append(mapping)
                    continue

                if pattern.severity == "hard":
                    idx = len(self._hard_mappings)
                    self._hard_mappings.append(mapping)
                    for lit in literals:
                        ac_idx = len(hard_literals)
                        hard_literals.append(lit)
                        mapping.ac_indices.append(ac_idx)
                        self._hard_lit_to_pat.setdefault(ac_idx, []).append(idx)
                else:
                    idx = len(self._soft_mappings)
                    self._soft_mappings.append(mapping)
                    for lit in literals:
                        ac_idx = len(soft_literals)
                        soft_literals.append(lit)
                        mapping.ac_indices.append(ac_idx)
                        self._soft_lit_to_pat.setdefault(ac_idx, []).append(idx)

        self._hard_automaton = AhoCorasick(hard_literals)
        self._soft_automaton = AhoCorasick(soft_literals)

    def scan(self, payload: str) -> list[ScanResult]:
        """Scan payload. Returns on first hard match (abort)."""
        # Phase 1: Hard
        hard_results = self._scan_automaton(
            payload, self._hard_automaton, self._hard_mappings, self._hard_lit_to_pat,
        )
        if hard_results:
            return [hard_results[0]]

        # Hard direct-regex
        for mapping in self._direct_patterns:
            if mapping.severity != "hard":
                continue
            result = self._confirm_regex(payload, mapping)
            if result:
                return [result]

        # Phase 2: Soft
        results: list[ScanResult] = []
        soft_results = self._scan_automaton(
            payload, self._soft_automaton, self._soft_mappings, self._soft_lit_to_pat,
        )
        results.extend(soft_results)

        for mapping in self._direct_patterns:
            if mapping.severity != "soft":
                continue
            result = self._confirm_regex(payload, mapping)
            if result:
                results.append(result)

        return results

    # -----------------------------------------------------------------------
    # Internal
    # -----------------------------------------------------------------------

    def _scan_automaton(
        self,
        payload: str,
        automaton: AhoCorasick,
        mappings: list[_PatternMapping],
        lit_to_pat: dict[int, list[int]],
    ) -> list[ScanResult]:
        results: list[ScanResult] = []
        ac_matches = automaton.search(payload)

        candidates: set[int] = set()
        for m in ac_matches:
            for idx in lit_to_pat.get(m.pattern_index, []):
                candidates.add(idx)

        for pat_idx in candidates:
            mapping = mappings[pat_idx]
            result = self._confirm_regex(payload, mapping)
            if result:
                results.append(result)

        return results

    @staticmethod
    def _confirm_regex(payload: str, mapping: _PatternMapping) -> Optional[ScanResult]:
        m = mapping.regex.search(payload)
        if not m:
            return None
        return ScanResult(
            scanner_id=mapping.scanner_id,
            pattern_id=mapping.pattern_id,
            severity=mapping.severity,
            match_start=m.start(),
            match_end=m.end(),
            match_text=m.group(0),
            scanner_label=mapping.scanner_label,
        )
